graph/nodes_edges.go

```diff
@@ -0,0 +1,209 @@
+// Copyright ©2018 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package graph
+
+// Iterator is a item iterator.
+type Iterator interface {
+	// Next advances the iterator and returns whether
+	// the next call to the item method will return a
+	// non-nil item.
+	//
+	// Next should be called prior to any call to the
+	// iterator's item retrieval method after the
+	// iterator has been obtained or reset.
+	//
+	// The order of iteration is implementation
+	// dependent.
+	Next() bool
+
+	// Len returns the number of items remaining in the
+	// iterator.
+	Len() int
+
+	// Reset returns the iterator to its start position.
+	Reset()
+}
+
+// Nodes is a Node iterator.
+type Nodes interface {
+	Iterator
+
+	// Node returns the current Node from the iterator.
+	Node() Node
+}
+
+// NodeSlicer wraps the NodeSlice method.
+type NodeSlicer interface {
+	// NodeSlice returns the set of nodes remaining
+	// to be iterated by a Nodes iterator.
+	// The holder of the iterator may arbitrarily
+	// change elements in the returned slice, but
+	// those changes may be reflected to other
+	// iterators.
+	NodeSlice() []Node
+}
+
+// NodesOf returns it.Len() nodes from it. If it is a NodeSlicer, the NodeSlice method
+// is used to obtain the nodes. It is safe to pass a nil Nodes to NodesOf.
+func NodesOf(it Nodes) []Node {
```

Comment:
Maybe this should be `func NodesOf(dst []Node, it Nodes) []Node` and append to `dst` and return it so that we can avoid allocations. Same for `EdgesOf` etc.

Reply:
That would make sense, and then passing nil for dst allocates a new slice. When if cap(dst) is less than the number of nodes in `it`?

Reply:
The behaviour would be `l := len(dst); dst = append(dst, make([]Node, it.Len())...)[:l]` followed by iterative addition of nodes for the non-slicer and just `dst = append(dst, it.NodeSlice()...)` for the slicer case.

This reduces the number of reallocations in the iterative case. It also gives a way to collect nodes from multiple `Nodes` values.